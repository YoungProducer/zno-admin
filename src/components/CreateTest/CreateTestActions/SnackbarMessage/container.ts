/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 18 February 2020
 *
 * Define main types and interfaces.
 * Define functions to get actions and varibales from redux store.
 * Export function which connect actions and store to 'TasksList' component.
 */

// External imports
import { connect } from 'react-redux';

// Application's imports
import { selectIsHaveErrorFields, selectIsHaveErrors } from 'store/selectors/createTest';
import { RootState } from 'store/slices';

/**
 * Props which component get from the parent.
 */
interface IOwnProps {
    message: React.ReactNode;
    key: string | number;
    haveErrors: boolean;
    haveErrorFields: boolean;
    closeSnackbar: (key: string | number) => void;
    toggleOpenTasksList: () => void;
}

/**
 * Props which component get from the redux store.
 */
interface IStateProps {
    // haveErrors: boolean;
    // haveErrorFields: boolean;
}

/**
 * Props(actions) which component can dispatch to store
 */
interface IDispatchProps {}

/**
 * Define type which describe all props which component get from the parent or from the store.
 */
export type TSnackbarMessageProps = IOwnProps & IStateProps & IDispatchProps;

/**
 * Define mapStateToProps function.
 * This function will connect variables from redux store to the component.
 */
const mapStateToProps = (state: RootState): IStateProps => ({
    // haveErrors: selectIsHaveErrors(state),
    // haveErrorFields: selectIsHaveErrorFields(state),
});

/**
 * Define mapDispatchToProps function.
 * This function will connect actions to the component.
 */
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
});

/**
 * Export function which connect actions and variables from the redux store to component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateToProps,
    mapDispatchToProps,
);
