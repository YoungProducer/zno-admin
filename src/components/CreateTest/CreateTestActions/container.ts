// Created by: Oleksandr Bezrukov
// Creation date: 3 February 2020

/**
 * Define all types and interface related to 'CreateTestAction' component.
 * Create function which connect state and actions to the component.
 */

// External imports
import { connect } from 'react-redux';

// Application's imports
import { createTestAction } from 'store/actionsCreators/createTest';
import { selectTaskSliceState } from 'store/selectors/task';
import { selectTasksList } from 'store/selectors/tasksList';
import {
    addTaskAction,
    TasksListSlice,
} from 'store/slices/tasksList';
import {
    clearTaskAction,
    TaskSlice,
} from 'store/slices/task';
import { RootState } from 'store/slices';

// Props which component get from parent
interface IOwnProps {}

// Props which component get from redux store
interface IStateProps {
    task: TaskSlice.State;
    tasksList: Partial<TasksListSlice.ExtendedTask>[];
}

// Props(actions) connected to the component
interface IDispatchProps {
    addTask: (payload: TasksListSlice.AddPayload) => void;
    clearTask: (revoke: boolean) => void;
    createTest: () => void;
}

// Define type of props for 'CreateTestActions' component which describe all props pushed to the component.
export type TCreateTestActionProps =
    IOwnProps
    & IStateProps
    & IDispatchProps;

/**
 * Define mapStateToProps function.
 * Connect store variables to component.
 */
const mapStateTopProps = (state: RootState): IStateProps => ({
    task: selectTaskSliceState(state),
    tasksList: selectTasksList(state),
});

/**
 * Define mapDispatchToProps function.
 * Connect actions to component.
 */
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    addTask: (payload: TasksListSlice.AddPayload) =>
        dispatch(addTaskAction(payload)),

    clearTask: (revoke: boolean) => dispatch(clearTaskAction(revoke)),

    createTest: () => dispatch(createTestAction()),
});

/**
 * Export function which connect state and actions to the component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateTopProps,
    mapDispatchToProps,
);
