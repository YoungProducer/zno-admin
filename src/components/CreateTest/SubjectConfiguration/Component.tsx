// Created by: Oleksandr Bezrukov
// Creation date: 30 January 2020

// Component to choose types of test and to give name for test.

// External imports
import React, { useState, useEffect } from 'react';
import Paper from '@material-ui/core/Paper';
import TextField from '@material-ui/core/TextField';
import CheckBox from '@material-ui/core/Checkbox';
import RadioGroup from '@material-ui/core/RadioGroup';
import Radio from '@material-ui/core/Radio';
import Collapse from '@material-ui/core/Collapse';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import FormControl from '@material-ui/core/FormControl';
import FormLabel from '@material-ui/core/FormLabel';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

// Application's imports
import AdvancedTextField from 'components/AdvancedTextField';

const useStyles = makeStyles((theme: Theme) => createStyles({
    paper: {
        padding: theme.spacing(2),
        height: '100%',
    },
    part: {
        marginBottom: theme.spacing(2),
    },
}));

const subjects = [
    'Математика',
    'Англійська мова',
];

export enum ETestTypes {
    'THEMES' = '0',
    'EXAM' = '1',
}

export enum EExamTypes {
    'TRAINING' = '0',
    'PREV_SESSIONS' = '1',
}

const Component = () => {
    const classes = useStyles({});

    const [subjectName, setSubjectName] = useState<string>('');

    const [withSubSubject, toggleWithSubSubject] = useState<boolean>(false);
    const [subSubjectName, setSubSubjectName] = useState<string>('');

    const [testType, setTestType] = useState<string>(ETestTypes.THEMES);
    const handleChangeTestType = (event: React.ChangeEvent<HTMLInputElement>) => {
        setTestType((event.target as HTMLInputElement).value);
    };

    const [examType, setExamType] = useState<string>(EExamTypes.TRAINING);
    const handleChangeExamType = (event: React.ChangeEvent<HTMLInputElement>) => {
        setExamType((event.target as HTMLInputElement).value);
    };

    const [themeName, setThemeName] = useState<string>('');

    useEffect(() => {
        setTestType(ETestTypes.THEMES);
    },        [withSubSubject]);

    return (
        <Paper
            elevation={3}
            className={classes.paper}
        >
            <div className={classes.part}>
                <AdvancedTextField
                    value={subjectName}
                    callback={setSubjectName}
                    list={subjects}
                    label='Назва предмету'
                    variant='standard'
                    color='secondary'
                />
            </div>
            <div className={classes.part}>
                <FormControlLabel
                    control={
                        <CheckBox
                            checked={withSubSubject}
                            onChange={event => toggleWithSubSubject(event.target.checked)}
                        />
                    }
                    label='Вибрати під-предмет'
                />
                <Collapse
                    in={withSubSubject}
                >
                    <AdvancedTextField
                        value={subSubjectName}
                        callback={setSubSubjectName}
                        list={[]}
                        label='Під-предмет'
                        variant='standard'
                        color='secondary'
                    />
                </Collapse>
            </div>
            <div className={classes.part}>
                <FormControl component='fieldset'>
                    <FormLabel
                        component='legend'
                        color='secondary'
                    >
                        Оберіть тип тесту
                    </FormLabel>
                    <RadioGroup value={testType} onChange={handleChangeTestType}>
                        <FormControlLabel
                            value={ETestTypes.THEMES}
                            control={<Radio color='secondary'/>}
                            label='Підготовка по темам'
                        />
                        <FormControlLabel
                            value={ETestTypes.EXAM}
                            control={<Radio color='secondary' disabled={withSubSubject}/>}
                            label='ЗНО'
                        />
                    </RadioGroup>
                </FormControl>
                <Collapse in={testType === ETestTypes.THEMES}>
                    <AdvancedTextField
                        value={themeName}
                        callback={setThemeName}
                        list={[]}
                        label='Назва теми'
                        color='secondary'
                        variant='standard'
                    />
                </Collapse>
            </div>
            <Collapse
                in={testType === ETestTypes.EXAM}
                className={classes.part}
            >
                <FormControl component='fieldset'>
                    <FormLabel
                        component='legend'
                        color='secondary'
                    >
                        Оберіть тип тесту
                    </FormLabel>
                    <RadioGroup value={examType} onChange={handleChangeExamType}>
                        <FormControlLabel
                            value={EExamTypes.TRAINING}
                            control={<Radio color='secondary'/>}
                            label='Тренувальні варінти ЗНО'
                        />
                        <FormControlLabel
                            value={EExamTypes.PREV_SESSIONS}
                            control={<Radio color='secondary' disabled={withSubSubject}/>}
                            label='Попередні сессії ЗНО'
                        />
                    </RadioGroup>
                </FormControl>
                <AdvancedTextField
                    value={themeName}
                    callback={setThemeName}
                    list={[]}
                    label='Назва'
                    color='secondary'
                    variant='standard'
                />
            </Collapse>
        </Paper>
    );
};

export default Component;
