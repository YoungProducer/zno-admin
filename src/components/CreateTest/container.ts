// Created by: Oleksandr Bezrukov
// Creation date: 3 February 2020

// Define main interfaces and types related to CreateTest component.
// Create function which connect state and actions to the component.

// External imports
import { connect } from 'react-redux';

// Application's imports
import {
    setTaskImageAction,
    setExplanationImageAction,
    deleteTaskImageAction,
    deleteExplanationImageAction,
} from 'store/slices/createTest';

// Props which component get from parent
interface IOwnProps {}

// Props(actions) which component can dispatch
interface IDispatchProps {
    setTaskImage: (file: File) => void;
    setExplanationImage: (file: File) => void;
    deleteTaskImage: () => void;
    deleteExplanationImage: () => void;
}

// Create type for CreateTest component which contain all props pushed to this component
export type TCreateTestProps = IOwnProps & IDispatchProps;

// Define mapDispatchToProps function
// Connect actions related to this component
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    setTaskImage: (file: File) => dispatch(setTaskImageAction(file)),
    setExplanationImage: (file: File) => dispatch(setExplanationImageAction(file)),
    deleteTaskImage: () => dispatch(deleteTaskImageAction()),
    deleteExplanationImage: () => dispatch(deleteExplanationImageAction()),
});

// Create function which connect state and actions to the component
export default connect<{}, IDispatchProps, IOwnProps>(
    null,
    mapDispatchToProps,
);
