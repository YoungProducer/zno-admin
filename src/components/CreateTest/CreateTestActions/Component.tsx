// Created by: Oleksandr Bezrukov
// Creation date: 3 February 2020

/**
 * Define styles and component.
 * Component related to main action of CreateTest: 'Add task' and 'Create test'.
 */

// External imports
import React from 'react';
import Button from '@material-ui/core/Button';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

// Application's imports
import { TCreateTestActionProps } from './container';

// Define and describe classes as hook
const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        marginTop: theme.spacing(2),
        width: '100%',
        display: 'flex',
        justifyContent: 'space-between',
        [theme.breakpoints.up('lg')]: {
            justifyContent: 'flex-start',
            '& button': {
                marginRight: theme.spacing(3),
            },
        },
    },
    button: {
        color: '#fff',
    },
}));

const Component = ({
    task,
    tasksList,
    addTask,
    clearTask,
    createTest,
}: TCreateTestActionProps) => {
    // Declare and define classes variable
    const classes = useStyles({});

    const handleAddTask = () => {
        addTask(task);
        clearTask(false);
    };

    return (
        <div className={classes.root}>
            <Button
                variant='contained'
                color='primary'
                disableElevation
                onClick={handleAddTask}
            >
                Додати завдання
            </Button>
            <Button
                variant='contained'
                color='primary'
                disableElevation
                data-testid='create-test'
                onClick={createTest}
            >
                Створити тест
            </Button>
        </div>
    );
};

export default Component;
