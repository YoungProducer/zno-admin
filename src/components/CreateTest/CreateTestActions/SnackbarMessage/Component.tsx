/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 18 February 2020
 *
 * Snackbar action which display large amount of info.
 */

// External imports
import React, { useState, forwardRef } from 'react';
import classNames from 'classnames';
import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import Typography from '@material-ui/core/Typography';
import Collapse from '@material-ui/core/Collapse';
import Paper from '@material-ui/core/Paper';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Divider from '@material-ui/core/Divider';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

// Application's imports
import { TSnackbarMessageProps } from './container';

// Define styles as hook
const useStyles = makeStyles((theme: Theme) => createStyles({
    card: {
        maxWidth: 400,
        minWidth: 344,
    },
    actionRoot: {
        padding: '8px 8px 8px 16px',
        backgroundColor: '#d32f2f',
        display: 'flex',
        justifyContent: 'space-between',
    },
    icons: {
        marginLeft: 'auto',
    },
    expand: {
        padding: '8px 8px',
        transform: 'rotate(0deg)',
        transition: theme.transitions.create('transform', {
            duration: theme.transitions.duration.shortest,
        }),
    },
    expandOpen: {
        transform: 'rotate(180deg)',
    },
    icon: {
        color: '#fff',
    },
    typography: {
        color: '#fff',
    },
    button: {
        minWidth: 130,
    },
    paper: {
        padding: theme.spacing(1),
    },
    contentTypography: {
        paddingTop: theme.spacing(0.5),
        paddingBottom: theme.spacing(0.5),
    },
}));

const Component = forwardRef(
({
    message,
    key,
    closeSnackbar,
    toggleOpenTasksList,
    hasTasksListError,
    hasSubjectConfigError,
}: TSnackbarMessageProps,
 ref,
) => {
    const classes = useStyles({});

    const [expanded, setExpanded] = useState<boolean>(false);

    const handleExpandClick = () => setExpanded(!expanded);
    const handleCloseClick = () => closeSnackbar(key);

    return (
        <Card className={classes.card} ref={ref}>
            <CardActions classes={{ root: classes.actionRoot }}>
                <Typography
                    variant='subtitle1'
                    className={classes.typography}
                >
                    {message}
                </Typography>
                <div className={classes.icons}>
                    <IconButton
                        size='small'
                        className={classNames(classes.expand, classes.icon, {
                            [classes.expandOpen]: !expanded,
                        })}
                        onClick={handleExpandClick}
                    >
                        <ExpandMoreIcon />
                    </IconButton>
                    <IconButton
                        size='small'
                        className={classes.icon}
                        onClick={handleCloseClick}
                    >
                        <CloseIcon />
                    </IconButton>
                </div>
            </CardActions>
            <Collapse in={expanded}>
                <Paper className={classes.paper}>
                    <List>
                        { hasSubjectConfigError && (
                            <ListItem divider>
                                <Typography className={classes.contentTypography}>
                                    Є незаповненні поля у налаштуваннях теста.
                                </Typography>
                            </ListItem>
                        )}
                        { hasTasksListError && (
                            <ListItem divider>
                                <Typography className={classes.contentTypography}>
                                    Деякі завдання заповнені неправильно або їх взагалі не створено.
                                </Typography>
                                <Button
                                    className={classes.button}
                                    onClick={() => {
                                        toggleOpenTasksList();
                                        handleCloseClick();
                                    }}
                                    color='primary'
                                    variant='contained'
                                    disableElevation
                                >
                                    Переглянути
                                </Button>
                            </ListItem>
                        )}
                    </List>
                </Paper>
            </Collapse>
        </Card>
    );
});

export default Component;
