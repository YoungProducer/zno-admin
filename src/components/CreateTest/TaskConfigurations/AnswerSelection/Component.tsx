// Created by: Oleksandr Bezrukov
// Creation date: 31 January 2020

// Component which allow to select right answer for different types of tasks.

// Externals imports
import React from 'react';

// Application's imports
import OneRightAnswer from './OneRightAnswer';
import RelationAnswer from './RelationAnswer';
import TextAnswer from './TextAnswer';
import { TAnswerSelectionProps } from './container';

const Component = ({
    taskType,
    ...store
}: TAnswerSelectionProps) => {

    return (
        <>
            {taskType === 'SINGLE' && (
                <OneRightAnswer {...store} />
            )}
            {taskType === 'RELATIONS' && (
                <RelationAnswer {...store} />
            )}
            {taskType === 'TEXT' && (
                <TextAnswer {...store} />
            )}
        </>
    );
};

export default Component;
