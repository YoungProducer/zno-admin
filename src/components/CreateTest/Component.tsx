// Created by: Oleksandr Bezrukov
// Creation date: 30 January 2020

// External imports
import React, { useState } from 'react';
import Paper from '@material-ui/core/Paper';
import Typography from '@material-ui/core/Typography';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

// Application's imports
import TaskConfigurations from './TaskConfigurations';
import UploadImages from './UploadImages';
import CreateTestActions from './CreateTestActions';
import { TCreateTestProps } from './container';

const useStyles = makeStyles((theme: Theme) => createStyles({
    root: {
        width: `100%`,
        padding: theme.spacing(2),
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
    },
    title: {
        marginBottom: theme.spacing(2),
    },
}));

const Component = () => {
    // Declare and define classes.
    const classes = useStyles({});

    return (
        <Paper
            elevation={3}
            className={classes.root}
        >
            <Typography
                variant='h4'
                color='secondary'
                className={classes.title}
            >
                Створення завдань та тесту
            </Typography>
            <TaskConfigurations />
            <UploadImages />
            <CreateTestActions />
        </Paper>
    );
};

export default Component;
