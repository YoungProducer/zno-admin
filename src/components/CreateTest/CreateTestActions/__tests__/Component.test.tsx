/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 13 February 2020
 *
 * Create test suites for CreateTestActions component.
 */

// Externail imports
import React from 'react';
import { shallow, mount, render } from 'enzyme';

// Application's imports
import Component from '../Component';
import { TCreateTestActionProps } from '../container';
import { EExamTypes, ETestTypes } from '../../../panels/SubjectConfigurationsPanel/Component';
import { ETaskType } from '../../../../store/slices/createTest/taskBuffer';

describe('Create test component', () => {
    const props: TCreateTestActionProps = {
        addTask: jest.fn(),
        clearTaskBuffer: jest.fn(),
        fetchCreateTest: jest.fn(),
        mainFields: {
            subjectName: 'foo',
            subSubjectName: '',
            testType: ETestTypes.THEMES,
            examType: EExamTypes.TRAINING,
            themeName: '',
        },
        taskBuffer: {
            answersAmount: 1,
            taskType: ETaskType.ONE_RIGHT,
            answer: 1,
            taskImage: null,
            explanationImage: null,
        },
        tasksList: [],
    };

    test('to match snapshot', () => {
        const root = render(<Component {...props}/>);

        expect(root).toMatchSnapshot();
    });
});
