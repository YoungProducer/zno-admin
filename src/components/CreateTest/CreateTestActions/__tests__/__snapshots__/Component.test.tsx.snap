// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create test component to match snapshot 1`] = `
<div
  class="makeStyles-root-1"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Додати завдання
    </span>
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Створити тест
    </span>
  </button>
</div>
`;
